hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+
+                            let sclh = ((ratio / (presc + 1)) - 3) / 3;
+                            let scll = 2 * (sclh + 1) - 1;
+
+                            let (sdadel, scldel) = if freq > 400.kHz() {
+                                // fast-mode plus
+                                let sdadel = 0;
+                                let scldel = i2cclk / 4_000_000 / (presc + 1) - 1;
+
+                                (sdadel, scldel)
+                            } else {
+                                // fast-mode
+                                let sdadel = i2cclk / 8_000_000 / (presc + 1);
+                                let scldel = i2cclk / 2_000_000 / (presc + 1) - 1;
+
+                                (sdadel, scldel)
+                            };
+
+                            (presc, scll, sclh, sdadel, scldel)
+                        } else {
+                            // standard-mode
+                            // here we pick SCLL = SCLH
+                            let presc = ratio / 514;
+
+                            let sclh = ((ratio / (presc + 1)) - 2) / 2;
+                            let scll = sclh;
+
+                            let sdadel = i2cclk / 2_000_000 / (presc + 1);
+                            let scldel = i2cclk / 800_000 / (presc + 1) - 1;
+
+                            (presc, scll, sclh, sdadel, scldel)
+                        };
+
+                        assert!(presc < 16);
+                        assert!(scldel < 16);
+                        assert!(sdadel < 16);
+                        let sclh = u8::try_from(sclh).unwrap();
+                        let scll = u8::try_from(scll).unwrap();
+
+                        // Configure for "fast mode" (400 KHz)
+                        // NOTE(write): writes all non-reserved bits.
+                        i2c.timingr.write(|w| {
+                            w.presc()
+                                .bits(presc as u8)
+                                .sdadel()
+                                .bits(sdadel as u8)
+                                .scldel()
+                                .bits(scldel as u8)
+                                .scll()
+                                .bits(scll)
+                                .sclh()
+                                .bits(sclh)
+                        });
+
+                        // Enable the peripheral
+                        i2c.cr1.modify(|_, w| w.pe().set_bit());
+
+                        Self { i2c, pins }
+                    }
+
+                /// Releases the I2C peripheral and associated pins
+                pub fn free(self) -> ($I2CX, (SCL, SDA)) {
+                    (self.i2c, self.pins)
+                }
+            }
+
+            impl<PINS> Read for $I2cX<PINS> {
+                type Error = Error;
+
+                fn read(&mut self, addr: u8, buffer: &mut [u8]) -> Result<(), Self::Error> {
+                    assert!(!buffer.is_empty());
```

Comment:
Would it make more sense to do this (is it valid)?
```rust
if buffer.is_empty() {
    return Ok(())
}
```

Smaller binary size (even with defmt), and one less panic location.

Also should this check exist on `write`?

Reply:
We are checking the opposite, no? i.e. it panics when the buffer is empty, and if it is not it continues.

Basically, we should not trigger any read operations if there is no place to store the bytes.

In the case of `write`, a 0-byte write is valid (the master writes the address and that's it). This is useful in devices without registers. We could argue that this is also applicable to `write_read` but that's oriented to longer transfers as far as I understand, so it makes no sense to deal with the 0-byte write edge case (which is why it also panics if any of the buffers is empty).

Reply:
Thanks for the explanation, I appreciate it!

The assert sounds good to me, can you add a comment for the panic condition then?